src/cir.rs

```diff
@@ -0,0 +1,451 @@
+// cir:  compact intermediate representation
+use crate::ast::Predicate;
+use crate::context::{Context, Match};
+use crate::interpreter::Execute;
+use crate::lir::{is_operator, LirInstruction, LirLogicalOperator, LirProgram, Translate};
+use crate::semantics::FieldCounter;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct CirProgram {
+    pub(crate) instructions: Vec<CirInstruction>,
+}
+
+impl CirProgram {
+    pub fn new() -> Self {
+        Self {
+            instructions: Vec::new(),
+        }
+    }
+}
+
+impl Default for CirProgram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum CirInstruction {
+    AndIns(AndIns),
+    OrIns(OrIns),
+    NotIns(NotIns),
+    Predicate(Predicate),
+}
+
+#[derive(Debug)]
+pub struct AndIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct OrIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct NotIns {
+    right: CirOperand,
+}
+
+#[derive(Debug, Clone)]
+pub enum CirOperand {
+    Index(usize),
+    Predicate(Predicate),
```

Comment:
Shall we move the `predicate` to the same level as logic operators for more balanced sizes of nodes? This way it also saves memory usage.

Reply:
yes, it is defined inside enum "CirInstruction".

Reply:
right, that could save a few bytes and it depends on the matcher pattern distribution. And, this will increase the memory access which degrades runtime performance.